contracts/src/v0.8/dev/automation/upkeeps/ProxyBalanceMonitor.sol

```diff
@@ -0,0 +1,287 @@
+
+  /**
+   * @param linkTokenAddress the LINK token address
+   * @param minBalance the minimum balance an aggregator can have before initiating a top up
+   * @param topUpAmount the amount of LINK to top up an aggregator with at once
+   */
+  constructor(address linkTokenAddress, uint256 minBalance, uint256 topUpAmount) ConfirmedOwner(msg.sender) {
+    require(linkTokenAddress != address(0));
+    require(minBalance > 0);
+    require(topUpAmount > 0);
+    LINK_TOKEN = IERC20(linkTokenAddress);
+    s_minBalance = minBalance;
+    s_topUpAmount = topUpAmount;
+  }
+
+  /**
+   * @notice Sets the list of subscriptions to watch and their funding parameters
+   * @param addresses the list of proxy addresses to watch
+   */
+  function setWatchList(address[] calldata addresses) external onlyOwner {
+    // first, remove all existing addresses from list
+    for (uint256 idx = s_watchList.length(); idx > 0; idx--) {
+      require(s_watchList.remove(s_watchList.at(idx - 1)));
+    }
+    // then set new addresses
+    for (uint256 idx = 0; idx < addresses.length; idx++) {
+      if (s_watchList.contains(addresses[idx])) {
+        revert DuplicateAddress(addresses[idx]);
+      }
+      if (addresses[idx] == address(0)) {
+        revert InvalidWatchList();
+      }
+      s_watchList.add(addresses[idx]);
+    }
+    emit WatchlistUpdated();
+  }
+
+  /**
+   * @notice Adds addresses to the watchlist without overwriting existing members
+   * @param addresses the list of proxy addresses to watch
+   */
+  function addToWatchList(address[] calldata addresses) external onlyOwner {
+    for (uint256 idx = 0; idx < addresses.length; idx++) {
+      if (s_watchList.contains(addresses[idx])) {
+        revert DuplicateAddress(addresses[idx]);
+      }
+      if (addresses[idx] == address(0)) {
+        revert InvalidWatchList();
+      }
+      s_watchList.add(addresses[idx]);
+    }
+    emit WatchlistUpdated();
+  }
+
+  /**
+   * @notice Gets a list of proxies that are underfunded, up to the MAX_PERFORM size
+   * @dev the function starts at a random index in the list to avoid biasing the first
+   * addresses in the list over latter ones.
+   * @dev the function will check at most MAX_CHECK proxies in a single call
+   * @dev the function returns a list with a max length of MAX_PERFORM
+   * @return list of proxy addresses whose aggregators are underfunded
+   */
+  function sampleUnderfundedAddresses() public view returns (address[] memory) {
+    uint256 numTargets = s_watchList.length();
+    uint256 numChecked = 0;
+    uint256 numToCheck = MAX_CHECK;
+    uint256 idx = uint256(blockhash(block.number - 1)) % numTargets; // start at random index, to distribute load
+    numToCheck = numTargets < MAX_CHECK ? numTargets : numTargets;
+    uint256 numFound = 0;
+    address[] memory proxiesToFund = new address[](MAX_PERFORM);
+    for (; numChecked < numToCheck; (idx, numChecked) = ((idx + 1) % numTargets, numChecked + 1)) {
```

Comment:
nit. can be potentially simplified by having a simple loop of `i` from `0` to `numToCheck-1`, and getting `s_watchList.at((idx+i)%numTargets)`